#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [34]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import median_absolute_error
from sklearn.metrics import explained_variance_score
from sklearn.metrics import max_error
import urllib.request
from PIL import Image

## Data

In [35]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/train 4.csv", index_col = 0)
test = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 4.csv", index_col = 0)


## Data exploration

In [36]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 912 entries, 0 to 911
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Ram               912 non-null    int64  
 1   Weight            912 non-null    float64
 2   Memory            912 non-null    int64  
 3   ScreenResolution  912 non-null    int64  
 4   Price_euros       912 non-null    float64
 5   Company           912 non-null    float64
 6   TypeName          912 non-null    float64
 7   Cpu               912 non-null    float64
 8   OpSys             912 non-null    float64
 9   Gpu               912 non-null    float64
dtypes: float64(7), int64(3)
memory usage: 78.4 KB


In [37]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 391 entries, 0 to 390
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Ram               391 non-null    int64  
 1   Weight            391 non-null    float64
 2   Memory            391 non-null    int64  
 3   ScreenResolution  391 non-null    int64  
 4   Company           391 non-null    float64
 5   TypeName          391 non-null    float64
 6   Cpu               391 non-null    float64
 7   Gpu               391 non-null    float64
 8   OpSys             391 non-null    float64
dtypes: float64(6), int64(3)
memory usage: 30.5 KB


In [38]:
train.head()

Ram  Weight  Memory  ScreenResolution  Price_euros      Company  \
0    4    2.40     500           1049088        387.0   560.810290   
1    8    1.23     256           5760000       1379.0  1180.813491   
2    8    3.00     256           2073600        854.0   560.810290   
3    8    1.26     512           2073600       1483.0  1086.775714   
4    8    2.80    1024           1440000        519.0  1079.256930   

      TypeName          Cpu        OpSys          Gpu  
0   788.550835   435.000000  1156.522676   446.861250  
1  1536.767801  1584.005601  1156.522676  1799.077143  
2   788.550835  1013.174710  1156.522676  1048.985000  
3   675.884211  1584.005601  1156.522676  1172.156294  
4   788.550835   405.031111  1156.522676   405.000000

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [39]:
X = train.drop("Price_euros", axis = 1)
y = train['Price_euros']
X.shape

(912, 9)

In [40]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [41]:
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

### 3. Asignar el modelo (vacío) a una variable


In [42]:
lm = LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [43]:
lm.fit(X, y)

LinearRegression()

In [44]:
coeff_df = pd.DataFrame(lm.coef_,
                        X.columns,
                        columns=['Coefficient'])
coeff_df.sort_values('Coefficient', ascending=False)

Coefficient
Ram                 44.152248
Gpu                  0.496602
OpSys                0.291895
Cpu                  0.238048
Company              0.144984
TypeName             0.133291
ScreenResolution     0.000069
Memory              -0.114268
Weight             -87.722469

In [45]:
predictions = lm.predict(X)
predictions

array([ 2.21549855e+02,  1.99067594e+03,  8.81246605e+02,  1.26292218e+03,
        3.77600228e+02,  1.42645663e+03,  8.59498854e+02,  8.92995210e+02,
        1.57194629e+03,  1.18219359e+03,  1.64630023e+03,  4.07151419e+02,
        1.30841195e+03,  1.23530382e+03,  6.61421658e+02,  1.05116166e+03,
        9.82301423e+02, -2.99313615e+01,  1.57151042e+03,  1.17674753e+03,
        2.04520683e+03,  2.12474588e+03,  1.13597439e+03,  1.06405431e+03,
        1.42008163e+03,  5.67008654e+02,  1.35831241e+03,  1.27589972e+03,
        1.13884986e+03,  7.75546893e+02,  1.28204029e+03,  1.07918505e+03,
        7.09558064e+02,  1.72725157e+03,  1.57054194e+03,  2.14692523e+03,
        1.11407470e+03,  8.14768993e+02,  3.38818326e+02,  1.60243900e+03,
        1.15975176e+02,  2.29957083e+03,  2.80192912e+03,  8.47516751e+01,
        1.03096285e+03,  1.16142052e+03,  4.87492883e+02,  1.33283356e+03,
        7.83116318e+02,  3.65757163e+02,  1.76889537e+03,  8.21277886e+02,
        1.68936778e+03,  

In [46]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error ,mean_squared_error, r2_score

print("MAE: ", mean_absolute_error(y, predictions))
print("MAPE: ", mean_absolute_percentage_error(y,predictions)) # Es el complemento del R2
print("MSE: ", mean_squared_error(y,predictions))
print("RMSE: ", np.sqrt(mean_squared_error(y,predictions))) #es el mas/menos en la misma unidad que el target
print("R2: ", r2_score(y,predictions))

MAE:  218.61534413064473
MAPE:  0.23425709192231825
MSE:  86934.64285844147
RMSE:  294.8468125288816
R2:  0.8126794005037801


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

In [47]:
X.head()

Ram  Weight  Memory  ScreenResolution      Company     TypeName  \
0    4    2.40     500           1049088   560.810290   788.550835   
1    8    1.23     256           5760000  1180.813491  1536.767801   
2    8    3.00     256           2073600   560.810290   788.550835   
3    8    1.26     512           2073600  1086.775714   675.884211   
4    8    2.80    1024           1440000  1079.256930   788.550835   

           Cpu        OpSys          Gpu  
0   435.000000  1156.522676   446.861250  
1  1584.005601  1156.522676  1799.077143  
2  1013.174710  1156.522676  1048.985000  
3  1584.005601  1156.522676  1172.156294  
4   405.031111  1156.522676   405.000000

In [48]:
# from sklearn.preprocessing import PolynomialFeatures
# poly_feats = PolynomialFeatures(degree = 3)
# poly_feats.fit(X)
# X_poly = poly_feats.transform(X)
# pd.DataFrame(X_poly)



In [49]:
# pol_reg = LinearRegression()
# pol_reg.fit(X_poly, y)


In [50]:
# predictions = pol_reg.predict(y)


# df_preds = pd.DataFrame(predictions, columns = ['predictions'])
# df_preds['real_value'] = pd.Series(y_test)

# df_preds

# from sklearn.metrics import r2_score
# r2_score(pol_reg.predict(X_test), y_test)

In [82]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

poly_feats = PolynomialFeatures(degree=1)
poly_feats.fit(X)
X_poly = poly_feats.fit_transform(X)

pol_reg = LinearRegression()
pol_reg.fit(X_poly, y)

test_data_poly = poly_feats.transform(test)
predictions = pol_reg.predict(test_data_poly)
df_preds = pd.DataFrame(predictions, columns = ['predictions'])
df_preds

c:\Users\Rafael Ortega\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names must be in the same order as they were in fit.

  warnings.warn(message, FutureWarning)


predictions
0    1186.663811
1     483.827059
2     399.809918
3    1116.974284
4     528.187692
..           ...
386  1905.293640
387  2124.492792
388   761.818631
389   346.942179
390  1800.283339

[391 rows x 1 columns]

### 2. Carga los datos de `test.csv` para predecir.

In [52]:
# X_pred = pd.read_csv("C:/Users/Rafael Ortega/Proyectos/Machine_Learning/test/test 2.csv", index_col = 0)
X_pred = test
X_pred.head()

Ram  Weight  Memory  ScreenResolution      Company     TypeName  \
0    8    1.25     256           2073600  1208.718750   788.550835   
1    4    1.40      32           2304000   312.200000   788.550835   
2    2    1.40      32           2073600   312.200000   788.550835   
3    8    2.65    1024           2073600  1180.813491  1706.815970   
4    2    1.10      32           1049088  1064.729423  1303.180120   

           Cpu          Gpu        OpSys  
0  1013.174710  1172.156294  1156.522676  
1   289.815000   307.117391  1156.522676  
2   289.815000   376.653182  1156.522676  
3  1013.174710  1249.338537  1156.522676  
4   317.258167   339.031600  1156.522676

In [53]:
X_pred.shape

(391, 9)

**¿Por qué da error?** 

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [54]:
X_pred.head()

Ram  Weight  Memory  ScreenResolution      Company     TypeName  \
0    8    1.25     256           2073600  1208.718750   788.550835   
1    4    1.40      32           2304000   312.200000   788.550835   
2    2    1.40      32           2073600   312.200000   788.550835   
3    8    2.65    1024           2073600  1180.813491  1706.815970   
4    2    1.10      32           1049088  1064.729423  1303.180120   

           Cpu          Gpu        OpSys  
0  1013.174710  1172.156294  1156.522676  
1   289.815000   307.117391  1156.522676  
2   289.815000   376.653182  1156.522676  
3  1013.174710  1249.338537  1156.522676  
4   317.258167   339.031600  1156.522676

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [55]:
predictions_submit = lm.predict(X_pred)
predictions_submit

c:\Users\Rafael Ortega\AppData\Local\Programs\Python\Python37\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names must be in the same order as they were in fit.

  warnings.warn(message, FutureWarning)


array([1186.6638112 ,  483.8270587 ,  399.8099182 , 1116.97428446,
        528.18769217,  660.4396616 ,  560.10062902, 1543.38416219,
       1613.24906682, 1373.61032562,  839.91442386, 1228.96655938,
       1253.16335331, 1013.36743607,  791.34732932, 1069.62174469,
       1302.60138441, 2213.78468675,  709.07117446,  691.74885534,
       1337.22183566,  847.86129568,  657.94864425, 1039.89705181,
        188.57949767, 1445.73758141,  853.53743296,  654.2706857 ,
        547.70497788, 1236.66748222, 1165.31599101, 1118.87018521,
        813.33178755,  925.2882229 , 1632.80489351, 1204.20830507,
       1055.14328126, 1387.53122233,  474.7619998 , 1688.98958863,
       1817.40816544, 1100.78226489,  272.72445821, 1949.78034592,
       1047.28197908, 2052.21837543, 1070.51982006, 1485.49719506,
       1005.3588727 , 1482.54242258, 1071.62632743,  306.19928929,
        855.9327837 , 1313.837487  ,  418.06855935,  923.07434146,
        833.995559  ,  255.53651038, 1161.08906075, 1164.73319

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [56]:
sample = pd.read_csv("data/sample_submission.csv")

In [57]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [58]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [84]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions})

In [78]:
submission.head()

laptop_ID  Price_euros
0        750  1119.370159
1        726   895.424436
2        633   744.810401
3        363  1018.937907
4        319   715.766724

In [85]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [83]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [86]:
chequeator(submission)

You're ready to submit!
